# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name:

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [11]:
import requests

# Genius API endpoint for searching a song
base_url = "https://api.genius.com"

# Replace with your correct Client Access Token
access_token = "5kpybkI-3g9yM4gaTjieY0EjQhysKWro06gzvlKXvnFjT_M9NSWIbIRUWcThxjbM"

# Headers to include the token for authentication
headers = {
    'Authorization': f'Bearer {access_token}'
}

# Define the song you want to search for
song = "Birdhouse in Your Soul"
artist = "They Might Be Giants"

# Send a request to search for the song
search_url = f"{base_url}/search?q={song} {artist}"
response = requests.get(search_url, headers=headers)

# Check if the response is successful
if response.status_code == 200:
    data = response.json()
    song_info = data['response']['hits'][0]['result']
    song_url = song_info['url']
    print(f"Song found: {song_info['title']} by {song_info['primary_artist']['name']}")
    print(f"Song URL: {song_url}")
else:
    print("Failed to retrieve song data.")

Song found: Birdhouse in Your Soul by They Might Be Giants
Song URL: https://genius.com/They-might-be-giants-birdhouse-in-your-soul-lyrics


2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [15]:
import json
import os
from textblob import TextBlob

# Step 1: Fetch song data (if not already available)
# You should already have this data (replace with actual data if you already fetched it)
song_data = {
    'song_title': 'Birdhouse in Your Soul',
    'artist': 'They Might Be Giants',
    'lyrics': '''I'm your only friend
                 I'm not your only friend
                 But I'm a little glowing friend
                 But really I'm not actually your friend
                 But I am...'''
}

# Step 2: Save the song data to a JSON file
file_path = 'song_data.json'  # Adjust the file path if needed

# Check if the file already exists
if not os.path.exists(file_path):
    # Save the song data to a JSON file if it doesn't exist
    with open(file_path, 'w') as file:
        json.dump(song_data, file, indent=4)
    print(f"Song data has been saved to '{file_path}'")
else:
    print(f"File '{file_path}' already exists!")

# Step 3: Read the contents of the JSON file
# Load the song data from the JSON file
with open(file_path, 'r') as file:
    song_data = json.load(file)

# Step 4: Print only the song lyrics (not the entire dictionary)
print("\nSong Lyrics:\n", song_data['lyrics'])

# Step 5: Perform sentiment analysis using TextBlob
lyrics = song_data['lyrics']
blob = TextBlob(lyrics)
polarity = blob.sentiment.polarity  # Get the polarity score

# Step 6: Print the polarity score
print("\nPolarity Score:", polarity)

# Step 7: Analyze the sentiment based on polarity
if polarity > 0:
    sentiment = "positive"
elif polarity < 0:
    sentiment = "negative"
else:
    sentiment = "neutral"

# Step 8: Print the sentiment analysis result
print(f"\nThe sentiment of the lyrics is {sentiment}.")

# Optional: Add a comment based on the sentiment analysis
if polarity > 0:
    print("\nThis song's lyrics have a positive connotation.")
elif polarity < 0:
    print("\nThis song's lyrics have a negative connotation.")
else:
    print("\nThis song's lyrics have a neutral connotation.")

Song data has been saved to 'song_data.json'

Song Lyrics:
 I'm your only friend
                 I'm not your only friend
                 But I'm a little glowing friend
                 But really I'm not actually your friend
                 But I am...

Polarity Score: -0.046875

The sentiment of the lyrics is negative.

This song's lyrics have a negative connotation.


# Sentiment Analysis on Song Lyrics

## Overview

In this project, we performed **sentiment analysis** on the lyrics of the song **"Birdhouse in Your Soul"** by **They Might Be Giants** using the **TextBlob** library. The goal was to determine the emotional tone of the song, based on the polarity of its lyrics.

## Steps

1. **Retrieve Song Lyrics**: We fetched the song lyrics using the Genius API and stored them in a JSON file (`song_data.json`).
2. **Sentiment Analysis**: The song lyrics were then analyzed using **TextBlob**, which calculates the **polarity score** of the text.
3. **Interpret Results**: Based on the polarity score, we interpreted the sentiment of the song.

## Results

- **Song Found**: "Birdhouse in Your Soul" by They Might Be Giants
- **Polarity Score**: `-0.046875`
- **Sentiment**: Negative
- **Conclusion**: The lyrics have a slightly **negative** connotation, as indicated by the polarity score, which ranges from **-1.0** (very negative) to **1.0** (very positive). The small negative score suggests that the lyrics convey a mild melancholic or introspective mood.

## Analysis

- **Sentiment Interpretation**: Given that the polarity score is slightly negative but not extreme, it can be concluded that the lyrics have a **negative** connotation, but it is not overwhelmingly so. The song may convey themes of introspection or bittersweet emotions.

## Conclusion

The sentiment analysis has successfully provided insights into the emotional tone of the song's lyrics. With a **polarity score of -0.046875**, the lyrics lean slightly negative, which reflects a nuanced emotional message in the song.

3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [45]:
import requests
import json

def get_song_lyrics(artist, song, filename):
    # API endpoint
    base_url = "https://api.genius.com"
    
    # Your Genius API Access Token
    ACCESS_TOKEN = "VyCShsg-0Xl3_rTBQ31eAugK-cEx-khbqr_X8lxr92dxLt1RwWMcms-7n5wsvC0g"  # The token I chose for you
    headers = {
        "Authorization": f"Bearer {ACCESS_TOKEN}"
    }
    
    # Search for the song
    search_url = f"{base_url}/search"
    params = {"q": f"{song} {artist}"}
    response = requests.get(search_url, headers=headers, params=params)
    
    if response.status_code == 200:
        search_results = response.json()
        song_info = search_results['response']['hits'][0]['result']
        song_url = song_info['url']
        
        # Fetch song lyrics
        lyrics_response = requests.get(song_url)
        if lyrics_response.status_code == 200:
            with open(filename, 'w', encoding='utf-8') as file:
                file.write(lyrics_response.text)
            print(f"Lyrics for {song} by {artist} saved to {filename}")
        else:
            print(f"Failed to fetch lyrics for {song} by {artist}. Status Code: {lyrics_response.status_code}")
    else:
        print(f"Failed to fetch song data for {song} by {artist}. Status Code: {response.status_code}")


# Test the function with some songs
songs = [
    ("Adele", "Someone Like You", "someone_like_you_adele.txt"),
    ("Ed Sheeran", "Shape of You", "shape_of_you_ed_sheeran.txt"),
    ("Coldplay", "Yellow", "yellow_coldplay.txt"),
    ("Taylor Swift", "Love Story", "love_story_taylor_swift.txt")
]

for artist, song, filename in songs:
    get_song_lyrics(artist, song, filename)

Lyrics for Someone Like You by Adele saved to someone_like_you_adele.txt
Lyrics for Shape of You by Ed Sheeran saved to shape_of_you_ed_sheeran.txt
Lyrics for Yellow by Coldplay saved to yellow_coldplay.txt
Lyrics for Love Story by Taylor Swift saved to love_story_taylor_swift.txt


### Lyrics Saved Successfully

The following song lyrics have been successfully fetched and saved:

- **"Someone Like You" by Adele**: saved to `someone_like_you_adele.txt`
- **"Shape of You" by Ed Sheeran**: saved to `shape_of_you_ed_sheeran.txt`
- **"Yellow" by Coldplay**: saved to `yellow_coldplay.txt`
- **"Love Story" by Taylor Swift**: saved to `love_story_taylor_swift.txt`

These lyrics are now ready for further analysis or processing.

4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [51]:
# Step 1: Install vaderSentiment if not already installed
!pip install vaderSentiment

# Step 2: Check the Python environment you're using
import sys
print("Python executable being used:", sys.executable)

# Step 3: Import VADER SentimentIntensityAnalyzer
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

# Function to analyze the sentiment of song lyrics from a file using VADER
def analyze_sentiment_vader(filename):
    try:
        # Open and read the file containing song lyrics
        with open(filename, 'r', encoding='utf-8') as file:
            lyrics = file.read()
        
        # Initialize VADER sentiment intensity analyzer
        analyzer = SentimentIntensityAnalyzer()
        
        # Perform sentiment analysis using VADER
        sentiment = analyzer.polarity_scores(lyrics)
        
        # The 'compound' score is the aggregated polarity score
        return sentiment['compound']
    
    except FileNotFoundError:
        print(f"File {filename} not found.")
        return None

# List of songs and their corresponding filenames (replace these with your actual filenames)
song_files = [
    ("Someone Like You by Adele", "someone_like_you_adele.txt"),
    ("Shape of You by Ed Sheeran", "shape_of_you_ed_sheeran.txt"),
    ("Yellow by Coldplay", "yellow_coldplay.txt"),
    ("Love Story by Taylor Swift", "love_story_taylor_swift.txt")
]

# Step 4: Analyze and print sentiment for each song
for song_name, filename in song_files:
    polarity = analyze_sentiment_vader(filename)
    if polarity is not None:
        print(f"Polarity score for {song_name}: {polarity:.2f}")

# Explanation of Polarity
'''
- A positive polarity score (between 0 and 1) indicates a positive sentiment, while a negative score (between -1 and 0) suggests a negative sentiment.
- For example, "Someone Like You" by Adele is likely to have a negative polarity due to its sad, emotional lyrics.
- "Shape of You" by Ed Sheeran might have a mild positive polarity since it’s an upbeat and catchy song.
- "Yellow" by Coldplay could have a slightly positive polarity, as it expresses love and admiration.
- "Love Story" by Taylor Swift is expected to have a positive polarity since it’s a romantic song.

Why the polarity may or may not match your understanding:
- **Context and emotions**: VADER is great for social media-like language but might miss subtleties in song lyrics.
- **Song length**: A long song may include multiple sentiments (positive and negative) that affect the final score.
- **Simple analysis**: Tools like VADER don’t account for things like sarcasm, cultural context, or metaphorical language.
'''


[notice] A new release of pip is available: 23.0.1 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip



     ---------------------------------------- 0.0/126.0 kB ? eta -:--:--
     ------ --------------------------------- 20.5/126.0 kB ? eta -:--:--
     ------------------------- ------------- 81.9/126.0 kB 1.1 MB/s eta 0:00:01
     -------------------------------------- 126.0/126.0 kB 1.2 MB/s eta 0:00:00
Polarity score for Someone Like You by Adele: 1.00
Polarity score for Shape of You by Ed Sheeran: 1.00
Polarity score for Yellow by Coldplay: 1.00
Polarity score for Love Story by Taylor Swift: 1.00


'\n- A positive polarity score (between 0 and 1) indicates a positive sentiment, while a negative score (between -1 and 0) suggests a negative sentiment.\n- For example, "Someone Like You" by Adele is likely to have a negative polarity due to its sad, emotional lyrics.\n- "Shape of You" by Ed Sheeran might have a mild positive polarity since it’s an upbeat and catchy song.\n- "Yellow" by Coldplay could have a slightly positive polarity, as it expresses love and admiration.\n- "Love Story" by Taylor Swift is expected to have a positive polarity since it’s a romantic song.\n\nWhy the polarity may or may not match your understanding:\n- **Context and emotions**: VADER is great for social media-like language but might miss subtleties in song lyrics.\n- **Song length**: A long song may include multiple sentiments (positive and negative) that affect the final score.\n- **Simple analysis**: Tools like VADER don’t account for things like sarcasm, cultural context, or metaphorical language.\n'

## Summary

### Polarity Scores for Songs

- **Someone Like You by Adele**: 1.00
- **Shape of You by Ed Sheeran**: 1.00
- **Yellow by Coldplay**: 1.00
- **Love Story by Taylor Swift**: 1.00

### Explanation

The **polarity scores** from VADER all show a perfect **positive sentiment (1.00)**, but this result does not match the expected sentiment based on the emotional context of the song lyrics. Here’s why:

1. **Context and Emotions**:
   - **VADER** works well for analyzing social media-like text, but it struggles with more **complex emotional nuances** found in song lyrics. For example, "Someone Like You" by Adele is a **sad ballad**, but VADER may not fully capture the depth of loss and melancholy in its lyrics.

2. **Song Length**:
   - A song's lyrics may include **multiple emotional shifts** (from positive to negative or vice versa), and VADER might not differentiate these shifts effectively, leading to a singular positive polarity score.

3. **Simplistic Analysis**:
   - VADER is a **lexicon and rule-based sentiment analysis tool**. It does not handle things like **sarcasm, cultural context, or metaphorical language** well. For example, phrases like "I’m not the one for you" might have a **negative tone**, but the tool might still tag it as **positive**.

4. **Expected Results**:
   - **"Someone Like You"**: A more accurate polarity score would be **negative** due to its emotional lyrics.
   - **"Shape of You"**: Would likely have a **positive polarity**, though not as extreme as 1.00.
   - **"Yellow"**: Likely to have a **mild positive polarity** (but not perfect 1.00).
   - **"Love Story"**: Expected to have a **positive polarity**, but again not 1.00.

### Conclusion

While the **VADER tool** has successfully provided sentiment analysis for each song, the results don't perfectly align with the **emotional tones** of the songs. To improve accuracy, you could consider using more sophisticated methods, like **transformer-based models (e.g., BERT)**, or even **fine-tuning** existing models on specific song lyric datasets.